pkg/agent/multicluster/stretched_networkpolicy_controller.go

```diff
@@ -0,0 +1,426 @@
+	defer klog.InfoS("Shutting down controller", "controller", stretchedNetworkPolicyControllerName)
+	cacheSyncs := []cache.InformerSynced{s.podListerSynced, s.namespaceListerSynced, s.LabelIdentityListerSynced}
+	if !cache.WaitForNamedCacheSync(stretchedNetworkPolicyControllerName, stopCh, cacheSyncs...) {
+		return
+	}
+	s.initLabelIDCache()
+	s.initPodClassifierFlow()
+	for i := 0; i < stretchedNetworkPolicyWorker; i++ {
+		go wait.Until(s.worker, time.Second, stopCh)
+	}
+	<-stopCh
+}
+
+func (s *StretchedNetworkPolicyController) initLabelIDCache() {
+	labelIdentities, err := s.labelIdentityLister.List(labels.Everything())
+	if err == nil {
+		s.lock.Lock()
+		defer s.lock.Unlock()
+		for _, labelIdentity := range labelIdentities {
+			s.labelIdentityCache[labelIdentity.Spec.Label] = labelIdentity.Spec.ID
+		}
+	}
+}
+
+func (s *StretchedNetworkPolicyController) initPodClassifierFlow() {
+	pods, err := s.podLister.List(labels.Everything())
+	if err != nil {
+		klog.ErrorS(err, "Can't get Pods")
+		return
+	}
+	for _, pod := range pods {
+		if pod.Spec.HostNetwork {
+			continue
+		}
+		podNS, err := s.namespaceLister.Get(pod.Namespace)
+		if err != nil {
+			klog.ErrorS(err, "Can't get the Namespace of a Pod", "name", pod.Name, "namespace", pod.Namespace)
+			return
+		}
+		normalizedLabel := multicluster.GetNormalizedLabel(podNS.Labels, pod.Labels, podNS.Name)
+		classifierFlowUpdate := s.getClassifierFlowUpdate(getPodReference(pod), normalizedLabel)
+		s.queue.Add(classifierFlowUpdate)
+	}
+}
+
+// worker is a long-running function that will continually call the processNextWorkItem
+// function in order to read and process a message on the workqueue.
+func (s *StretchedNetworkPolicyController) worker() {
+	for s.processNextWorkItem() {
+	}
+}
+
+func (s *StretchedNetworkPolicyController) processNextWorkItem() bool {
+	obj, quit := s.queue.Get()
+	if quit {
+		return false
+	}
+	defer s.queue.Done(obj)
+
+	if c, ok := obj.(*classifierFlowInfo); !ok {
+		s.queue.Forget(obj)
+		klog.Errorf("Expected type 'classifierFlowInfo' in work queue but got object", "object", obj)
+	} else if err := s.syncPodClassifierFlow(c); err == nil {
+		s.queue.Forget(c)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		s.queue.AddRateLimited(c)
+		klog.ErrorS(err, "Error syncing classifierFlowInfo, requeuing", "name", c.PodRef.Name, "namespace", c.PodRef.Namespace, "labelID", c.LabelID)
+	}
+	return true
+}
+
+// syncPodClassifierFlow will update labelToPods and podToLabel and update
+// PodClassifierFlow with its LabelIdentity, if available, or UnknownLabelIdentity.
+func (s *StretchedNetworkPolicyController) syncPodClassifierFlow(classifierFlowUpdate *classifierFlowInfo) error {
+	containerConfigs := s.interfaceStore.GetContainerInterfacesByPod(classifierFlowUpdate.PodRef.Name, classifierFlowUpdate.PodRef.Namespace)
+	if len(containerConfigs) == 0 {
+		return fmt.Errorf("pod container config not found")
+	}
```

Comment:
would it cause many errors in production as a Pod could be enqueued by its update event, but it may not be processed by CNI server yet.

Reply:
If a Pod "update" event needs another round CNI server process, it is actually a “delete + create” and won't fall into update handler. The create func is triggered by CNI add events which ensures the `containerConfig` is ready when enqueued. Let me know if I misunderstand anything.

Reply:
doesn't processPodUpdate enqueue pods as well? what if a real Pod update event is received?

Reply:
processPodUpdate enqueue pods, but containerConfigs doesn't change?

Reply:
By which code it's guaranteed CNIServer must have processed the Pod and updated `interfaceStore` when this controller receives the Pod's update event? This is also my first comment: "a Pod could be enqueued by its update event, but it may not be processed by CNI server yet"

Reply:
I see your point. Based on the offline discussion, changed from `return error` to `info log`.